frontend/src/components/core/EnterNickname.tsx

```diff
@@ -0,0 +1,92 @@
+import React, { useState, useEffect } from 'react';
+import ErrorMessage from './Error';
+import Loading from './Loading';
+import { LargeText, Text } from './Text';
+import { LargeCenterInputText, LargeInputButton } from './Input';
+import { ErrorResponse } from '../../api/Error';
+import { isValidNickname } from '../../api/Socket';
+
+type EnterNicknameProps = {
+  enterNicknameHeaderText: string,
+  enterNicknameAction: (nickname: string) => Promise<undefined>;
+}
+
+export default function EnterNicknamePage(props: EnterNicknameProps) {
+  // Grab props variables.
+  const {
+    enterNicknameHeaderText, enterNicknameAction,
+  } = props;
+
+  // Hold error text.
+  const [error, setError] = useState('');
+
+  // Hold loading boolean, triggered upon entering nickname.
+  const [loading, setLoading] = useState(false);
+
+  // Variable to hold the user's current nickname input.
+  const [nickname, setNickname] = useState('');
+
+  /**
+   * The nickname is valid if it is non-empty, has no spaces, and
+   * is <= 16 characters. This is updated whenever the nickname changes.
+   */
+  const [validNickname, setValidNickname] = useState(false);
+  useEffect(() => {
+    setValidNickname(isValidNickname(nickname));
+  }, [nickname]);
+
+  // Variable to hold whether the user is focused on the text input field.
+  const [focusInput, setFocusInput] = useState(false);
+
+  return (
+    <div>
+      <LargeText>
+        {enterNicknameHeaderText}
+      </LargeText>
+      <LargeCenterInputText
+        placeholder="Your nickname"
+        onChange={(event: React.ChangeEvent<HTMLInputElement>) => {
+          setNickname(event.target.value);
+        }}
+        onFocus={() => {
+          setFocusInput(true);
+        }}
+        onBlur={() => {
+          setFocusInput(false);
+        }}
+        onKeyPress={(event) => {
+          if (event.key === 'Enter' && validNickname) {
+            setLoading(true);
+            enterNicknameAction(nickname).then(() => {
+              setLoading(false);
+            }).catch((err: ErrorResponse) => {
+              setLoading(false);
+              setError(err.message);
+            });
+          }
+        }}
+      />
+      <LargeInputButton
+        onClick={() => {
+          setLoading(true);
+          enterNicknameAction(nickname).then(() => {
+            setLoading(false);
+          }).catch((err: ErrorResponse) => {
+            setLoading(false);
+            setError(err.message);
+          });
```

Comment:
Would it be better to extract these lines into a function to avoid repeating them? I think that's what you did in `addUsersToWaitingRoom()`.

Reply:
I think that makes sense, I'll update this and makes sure it works -- thanks!

Reply:
Updated 👍